### PR TITLE
Optimize RoundRobinLoadBalance

### PR DESCRIPTION
**Pull request by gudegg**
## What is the purpose of the change

fix:https://github.com/apache/incubator-dubbo/issues/2578
## Brief changelog
Optimize RoundRobinLoadBalance

## Verifying this change

Follow this checklist to help us incorporate your contribution quickly and easily:

- [x] Make sure there is a [GITHUB_issue](https://github.com/apache/incubator-dubbo/issues) filed for the change (usually before you start working on it). Trivial changes like typos do not require a GITHUB issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
- [ ] Format the pull request title like `[Dubbo-XXX] Fix UnknownException when host config not exist #XXX`. Each commit in the pull request should have a meaningful subject line and body.
- [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [ ] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in [test module](https://github.com/apache/incubator-dubbo/tree/master/dubbo-test).
- [ ] Run `mvn clean install -DskipTests` & `mvn clean test-compile failsafe:integration-test` to make sure unit-test and integration-test pass.
- [ ] If this contribution is large, please follow the [Software Donation Guide](https://github.com/apache/incubator-dubbo/wiki/Software-donation-guide).

### CODE REVIEW

**Comment by carryxyh** [comment]
@kimmking 
Have a look. :)

File: dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/loadbalance/RoundRobinLoadBalance.java [hunk 1]
```diff
@@ -77,6 +77,6 @@
             }
         }
         // Round robin
-        return invokers.get(sequence.incrementAndGet() % length);
+        return invokers.get(sequence.getAndIncrement() % length);
```
Comment (kimmking) [inline dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/loadbalance/RoundRobinLoadBalance.java]: good！！

**Comment by kimmking** [comment]
@carryxyh  it's okay and works well.
I send a mail to describe it clear and shortly.
I will be merge it soon and then add some test cases for the new algorithm with performance case.

**Comment by gudegg** [comment]
@kimmking Have a look! I made a little change.
algorithm: Only need invokers#size loop at most.
qq20181005-125137 2x

File: dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/loadbalance/RoundRobinLoadBalance.java [hunk 2]
```diff
@@ -21,83 +21,64 @@
...
+
     @Override
     protected <T> Invoker<T> doSelect(List<Invoker<T>> invokers, URL url, Invocation invocation) {
         String key = invokers.get(0).getUrl().getServiceKey() + "." + invocation.getMethodName();
         int length = invokers.size(); // Number of invokers
         int maxWeight = 0; // The maximum weight
         int minWeight = Integer.MAX_VALUE; // The minimum weight
-        final LinkedHashMap<Invoker<T>, IntegerWrapper> invokerToWeightMap = new LinkedHashMap<Invoker<T>, IntegerWrapper>();
-        int weightSum = 0;
+        final List<Invoker<T>> invokerToWeightList = new ArrayList<>();
```
Comment (ralf0131) [inline dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/loadbalance/RoundRobinLoadBalance.java]: The name `invokerToWeightList` can be improved. I think `weightedInvokerList` or `nonZeroWeightedInvokers`  should be better. :)

**Comment by ralf0131** [comment]
This pull request looks good to me. Nice work! 
The time complexity is O(n), where n is the number of invokers, which is a significant improvement to the previous one, which is O(n*w), where n is the number of invokers, and w is the max weight of the invokers.
I'd prefer to have more unit tests for this algorithm.

File: dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/loadbalance/RoundRobinLoadBalance.java [hunk 3]
```diff
@@ -21,83 +21,64 @@
...
+                indexSeqs.putIfAbsent(key, new AtomicPositiveInteger(-1));
+                indexSeq = indexSeqs.get(key);
+            }
+            length = nonZeroWeightedInvokers.size();
+            while (true) {
+                int index = indexSeq.incrementAndGet() % length;
+                int currentWeight;
+                if (index == 0) {
+                    currentWeight = sequence.incrementAndGet() % maxWeight;
+                }else {
```
Comment (detailyang) [inline dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/loadbalance/RoundRobinLoadBalance.java]: space need:)

**Pull request merged**